File: CsharpTDD/CsharpTDD/Bank.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace CsharpTDD
{
    internal class Bank
    {
        private Dictionary<Pair, int> rates = new Dictionary<Pair, int>();

        public Money Reduce(IExpression source, string to)
        {
            return source.Reduce(this, to);
        }

        public int rate(string from, string to)
        {
            if (from.Equals(to)) return 1;
            return rates[new Pair(from, to)];
        }

        public void AddRate(string from, string to, int rate)
        {
            rates.Add(new Pair(from, to), rate);
        }
    }
}

File: JavaTdd/src/main/java/com/sangkon/Bank.java
package com.sangkon;

import java.util.Hashtable;

public class Bank {

    private Hashtable rates = new Hashtable();


    public void addRate(String from, String to, int rate) {
        rates.put(new Pair(from, to), rate);
    }

    public int rate(String from, String to) {
        if (from.equals(to)) return 1;
        int rate = (int) rates.get(new Pair(from, to));
        return rate;
    }

    public Money reduce(Expression source, String to) {
        return source.reduce(this, to);
    }

}
